es6-shim.js

```diff
@@ -387,20 +467,32 @@
     }
   };
 
-  var emulateES6construct = function (o) {
-    if (!ES.TypeIsObject(o)) { throw new TypeError('bad object'); }
-    var object = o;
-    // es5 approximation to es6 subclass semantics: in es6, 'new Foo'
-    // would invoke Foo.@@species to allocation/initialize the new object.
-    // In es5 we just get the plain object.  So if we detect an
-    // uninitialized object, invoke o.constructor.@@species
-    if (!object._es6construct) {
-      if (object.constructor && ES.IsCallable(object.constructor[symbolSpecies])) {
-        object = object.constructor[symbolSpecies](object);
-      }
-      defineProperties(object, { _es6construct: true });
+  var emulateES6construct = function (o, defaultNewTarget, defaultProto, slots) {
+    // This is an es5 approximation to es6 construct semantics.  in es6,
+    // 'new Foo' invokes Foo.[[Construct]] which (for almost all objects)
+    // just sets the internal variable NewTarget (in es6 syntax `new.target`)
+    // to Foo and then returns Foo().
+
+    // Many ES6 object then have constructors of the form:
+    // 1. If NewTarget is undefined, throw a TypeError exception
+    // 2. Let xxx by OrdinaryCreateFromConstructor(NewTarget, yyy, zzz)
+
+    // So we're going to emulate those first two steps.
+    if (!ES.TypeIsObject(o)) {
+      throw new TypeError('Constructor requires `new`: ' + defaultNewTarget.name);
     }
-    return object;
+    var proto = defaultNewTarget.prototype;
+    if (!ES.TypeIsObject(proto)) {
+      proto = defaultProto;
+    }
+    o = create(proto);
+    for (var name in slots) {
+      if (_hasOwnProperty(slots, name)) {
```

Comment:
I expect this to be very hot code, since it will be executed every time we create a Map, Set, or Promise.  I've gotten complaints that our implementations were slow to construct in the past.  The `for...in` loop ought to be optimized better than `Object.keys(...).forEach(function() { .... })` (which requires the creation of an extra array and function environment).  We don't really need to worry about obscure bugs because the `slots` are always coming from our code, this parameter isn't exposed to users of the shim.